File: backend/services/polyratings.py
"""
PolyRatings integration.

Fetches the full professor list from professors.all once per CACHE_TTL_SECONDS,
builds a course-number index, and returns filtered results on demand.

PolyRatings still stores quarter-system course numbers (CSC 101, etc.) for
all historical ratings.  We search by both the semester course number AND every
quarter equivalent so students see results regardless of which system was used.
"""

import re
import time
import httpx
from dataclasses import dataclass
from data.flowcharts import FLOWCHARTS

API_BASE   = "https://api-prod.polyratings.org"
PR_BASE    = "https://polyratings.dev"
CACHE_TTL  = 6 * 60 * 60  # 6 hours


@dataclass
class Professor:
    name: str
    overall_score: float
    num_ratings: int
    polyratings_url: str


# ── helpers ──────────────────────────────────────────────────────────────────

def _norm(course_number: str) -> str:
    """Normalise for comparison: upper-case, single space, strip."""
    parts = course_number.upper().split()
    return " ".join(parts)


_COURSE_RE = re.compile(r"(?P<dept>[A-Z]{2,5})\s*(?P<num>\d{3,4}[A-Z]?)")


def _extract_course_numbers(text: str) -> set[str]:
    return {f"{match.group('dept')} {match.group('num')}" for match in _COURSE_RE.finditer(text.upper())}


_STATIC_QUARTER_EQUIVALENTS: dict[str, set[str]] = {
    "CSC 3001": {"CSC 364", "CSC 464"},
    "CSC 3203": {"CSC 323"},
    "CSC 3445": {"CSC 445"},
    "CSC 3660": {"CSC 365"},
    "CSC 4160": {"CSC 402"},
    "CSC 4161": {"CSC 405", "CSC 406", "CSC 491", "CSC 492"},
    "CSC 4460": {"CSC 491", "CSC 492"},
    "CSC 4461": {"CSC 491", "CSC 492"},
    "CSC 4553": {"CSC 453", "CPE 453"},
    "CSC 4669": {"CSC 469"},
    "CPE 1000": {"CPE 100"},
    "CPE 2301": {"CPE 133"},
    "CPE 4220": {"CPE 422"},
    "CPE 4280": {"CPE 426"},
    "CPE 4669": {"CPE 469"},
    "DATA 4460": {"CSC 491", "CSC 492"},
    "ENGR 4464": {"CSC 491", "CSC 492"},
    "MATH 2031": {"MATH 248"},
}


def _build_index(professors: list[dict]) -> dict[str, list[dict]]:
    index: dict[str, list[dict]] = {}
    for prof in professors:
        for course in prof.get("courses", []):
            keys = {_norm(course), *{_norm(num) for num in _extract_course_numbers(course)}}
            for key in keys:
                index.setdefault(key, []).append(prof)
    return index


def _fetch_all() -> list[dict]:
    resp = httpx.get(f"{API_BASE}/professors.all", timeout=15)
    resp.raise_for_status()
    payload = resp.json()
    # tRPC shape: {"result": {"data": [...]}}
    return payload["result"]["data"]


# ── build reverse lookup: semester course_number -> quarter equivalents ───────

def _build_quarter_lookup() -> dict[str, set[str]]:
    """Map every semester course number to its known quarter equivalents (+ itself)."""
    lookup: dict[str, set[str]] = {}
    for fc in FLOWCHARTS.values():
        for course in fc["courses"]:
            nums: set[str] = {_norm(course["course_number"])}
            for q in course["quarter_equivalents"]:
                nums.add(_norm(q))
            for n in nums:
                lookup.setdefault(n, set()).update(nums)

    for semester, quarters in _STATIC_QUARTER_EQUIVALENTS.items():
        nums = {_norm(semester), *{_norm(q) for q in quarters}}
        for n in nums:
            lookup.setdefault(n, set()).update(nums)
    return lookup


_QUARTER_LOOKUP: dict[str, set[str]] = _build_quarter_lookup()


def _derive_quarter_candidates(normed: str) -> set[str]:
    """
    Generate conservative quarter-number candidates for semester courses.
    Lower-division examples often prepend a leading digit to the old number
    (COMS 1101 -> COMS 101). Upper-division semester courses that end in 0
    often append that 0 to the old 3-digit course (CSC 4710 -> CSC 471).
    """
    candidates: set[str] = set()
    parts = normed.split()
    if len(parts) == 2:
        dept, code = parts
        if len(code) == 4 and code.isdigit():
            stripped = str(int(code[1:]))  # "0101" -> "101", preserves "101"
            candidates.add(f"{dept} {stripped}")
            if int(code) >= 3000 and code.endswith("0"):
                candidates.add(f"{dept} {code[:3]}")
    return candidates


# ── cache class ───────────────────────────────────────────────────────────────

class PolyRatingsCache:
    """Thread-safe-enough in-memory TTL cache for the PolyRatings professor list."""

    def __init__(self) -> None:
        self._professors: list[dict] | None = None
        self._ts: float = 0.0
        self._index: dict[str, list[dict]] = {}

    def reset(self) -> None:
        """Clear cached state. Used by tests."""
        self._professors = None
        self._ts = 0.0
        self._index = {}

    def _ensure(self) -> None:
        if self._professors is not None and (time.time() - self._ts) < CACHE_TTL:
            return
        self._professors = _fetch_all()
        self._index = _build_index(self._professors)
        self._ts = time.time()

    def warm(self) -> None:
        """Pre-warm the cache. Safe to call from a background thread."""
        self._ensure()

    def get_professors_for_course(self, course_number: str) -> list[Professor]:
        """
        Return professors who have taught this course, sorted by numEvals desc.
        Searches the semester number, all FLOWCHARTS-based quarter equivalents, and
        a heuristic-derived quarter candidate (strip leading digit of 4-digit codes).
        """
        try:
            self._ensure()
        except Exception:
            return []

        normed = _norm(course_number)
        requested_numbers = {_norm(num) for num in _extract_course_numbers(course_number)} or {normed}
        search_keys: set[str] = set()
        for requested in requested_numbers:
            search_keys.add(requested)
            search_keys.update(_QUARTER_LOOKUP.get(requested, set()))
            search_keys.update(_derive_quarter_candidates(requested))

        seen_ids: set[str] = set()
        matched: list[dict] = []
        for key in search_keys:
            for prof in self._index.get(key, []):
                if prof["id"] not in seen_ids:
                    seen_ids.add(prof["id"])
                    matched.append(prof)

        matched.sort(key=lambda p: (p.get("numEvals", 0), p.get("overallRating", 0)), reverse=True)

        return [
            Professor(
                name=f"{p['firstName']} {p['lastName']}",
                overall_score=round(p.get("overallRating", 0.0), 2),
                num_ratings=p.get("numEvals", 0),
                polyratings_url=f"{PR_BASE}/professor/{p['id']}",
            )
            for p in matched
            if p.get("numEvals", 0) > 0
        ]


# ── module singleton + dependency ─────────────────────────────────────────────

_cache = PolyRatingsCache()


def get_polyratings_cache() -> PolyRatingsCache:
    """FastAPI dependency — returns the shared PolyRatingsCache singleton."""
    return _cache


# ── backwards-compatible free functions ───────────────────────────────────────

def warm_cache() -> None:
    """Pre-warm the PolyRatings cache. Intended to be called from the app lifespan."""
    _cache.warm()


def get_professors_for_course(course_number: str) -> list[Professor]:
    """Module-level shim kept for direct callers outside of FastAPI routes."""
    return _cache.get_professors_for_course(course_number)

File: frontend/lib/polyratings.ts
import type { Professor } from "./types";

type RawPolyRatingsProfessor = {
  id: string;
  firstName: string;
  lastName: string;
  courses?: string[];
  overallRating?: number;
  numEvals?: number;
};

type PolyRatingsPayload = {
  result?: {
    data?: RawPolyRatingsProfessor[];
  };
};

const POLYRATINGS_API_BASE = "https://api-prod.polyratings.org";
const POLYRATINGS_PROFILE_BASE = "https://polyratings.dev";

const COURSE_RE = /([A-Z]{2,5})\s*(\d{3,4}[A-Z]?)/g;
const DEPT_ALIASES: Record<string, string[]> = {
  EGL: ["ENGL"],
};

let professorCache: Promise<RawPolyRatingsProfessor[]> | null = null;

function normCourse(courseNumber: string): string {
  return courseNumber.toUpperCase().trim().replace(/\s+/g, " ");
}

function extractCourseNumbers(text: string): Set<string> {
  const courses = new Set<string>();
  for (const match of text.toUpperCase().matchAll(COURSE_RE)) {
    const dept = match[1];
    const num = match[2];
    if (dept && num) courses.add(`${dept} ${num}`);
  }
  return courses;
}

function departmentVariants(dept: string): Set<string> {
  return new Set([dept, ...(DEPT_ALIASES[dept] ?? [])]);
}

function deriveQuarterCandidates(courseNumber: string): Set<string> {
  const candidates = new Set<string>();
  const [dept, code] = normCourse(courseNumber).split(/\s+/);
  if (!dept || !code || !/^\d{4}$/.test(code)) return candidates;

  for (const deptVariant of departmentVariants(dept)) {
    candidates.add(`${deptVariant} ${String(Number(code.slice(1)))}`);
    if (Number(code) >= 3000 && code.endsWith("0")) {
      candidates.add(`${deptVariant} ${code.slice(0, 3)}`);
    }
  }
  return candidates;
}

function searchKeysForCourse(courseNumber: string): Set<string> {
  const requested = extractCourseNumbers(courseNumber);
  if (requested.size === 0) requested.add(normCourse(courseNumber));

  const keys = new Set<string>();
  for (const course of requested) {
    const [dept, code] = normCourse(course).split(/\s+/);
    const depts = dept ? departmentVariants(dept) : new Set<string>();
    for (const deptVariant of depts) {
      if (code) keys.add(`${deptVariant} ${code}`);
    }
    for (const candidate of deriveQuarterCandidates(course)) {
      keys.add(candidate);
    }
  }
  return keys;
}

function indexKeysForPolyRatingsCourse(course: string): Set<string> {
  const keys = new Set([normCourse(course)]);
  for (const extracted of extractCourseNumbers(course)) {
    keys.add(normCourse(extracted));
  }
  return keys;
}

function buildIndex(professors: RawPolyRatingsProfessor[]): Map<string, RawPolyRatingsProfessor[]> {
  const index = new Map<string, RawPolyRatingsProfessor[]>();
  for (const professor of professors) {
    for (const course of professor.courses ?? []) {
      for (const key of indexKeysForPolyRatingsCourse(course)) {
        index.set(key, [...(index.get(key) ?? []), professor]);
      }
    }
  }
  return index;
}

async function fetchAllProfessors(): Promise<RawPolyRatingsProfessor[]> {
  const res = await fetch(`${POLYRATINGS_API_BASE}/professors.all`);
  if (!res.ok) return [];
  const payload = (await res.json()) as PolyRatingsPayload;
  return payload.result?.data ?? [];
}

export function resetPolyRatingsCacheForTests(): void {
  professorCache = null;
}

export async function getPolyRatingsProfessorsForCourse(courseNumber: string): Promise<Professor[]> {
  professorCache ??= fetchAllProfessors().catch((error) => {
    professorCache = null;
    throw error;
  });

  const allProfessors = await professorCache;
  const index = buildIndex(allProfessors);
  const searchKeys = searchKeysForCourse(courseNumber);
  const seen = new Set<string>();
  const matched: RawPolyRatingsProfessor[] = [];

  for (const key of searchKeys) {
    for (const professor of index.get(key) ?? []) {
      if (seen.has(professor.id)) continue;
      seen.add(professor.id);
      matched.push(professor);
    }
  }

  return matched
    .filter((professor) => (professor.numEvals ?? 0) > 0)
    .sort((a, b) => {
      const evalDiff = (b.numEvals ?? 0) - (a.numEvals ?? 0);
      if (evalDiff !== 0) return evalDiff;
      return (b.overallRating ?? 0) - (a.overallRating ?? 0);
    })
    .map((professor) => ({
      name: `${professor.firstName} ${professor.lastName}`,
      overall_score: Number((professor.overallRating ?? 0).toFixed(2)),
      num_ratings: professor.numEvals ?? 0,
      polyratings_url: `${POLYRATINGS_PROFILE_BASE}/professor/${professor.id}`,
    }));
}
